add aliases for `\<letter>` interactive client commands

pmed opened the pull request:
This is another way to support `\<letter>` commands like:
  * `\d` aka `SHOW TABLES`
  * `\l` aka `SHOW DATABASES`
  * `\c` aka `USE`

It has been implemented as `input` string prefix replacement with an aliased command
inside of the loop of `ClientBase::runInteractive()` in order to allow
command parameters, like `\c db_name`.

Additionally this  allow us to use all the command parameters:

> :) \l like 'system'\G
>
> SHOW DATABASES LIKE 'system'
>
> Query id: 26edd1f6-1695-4be2-baac-9fa30f7d820d
>
> Row 1:
> ──────
> name: system


Automated testing seems not easy to do, since the aliases work only in interactive client mode.

See issue #9339


Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
added \l, \d, \c aliases like in MySQL


Detailed description / Documentation draft:
The following \<letter> commands are available:

  - `\d` as alias for `SHOW TABLES`
  - `\l` as alias for `SHOW DATABASES`
  - `\c` as alias for `USE db_name`
 

> By adding documentation, you'll allow users to try your new feature immediately, not when someone else will have time to document it later. Documentation is necessary for all features that affect user experience in any way. You can add brief documentation draft above, or add documentation right into your patch as Markdown files in [docs](https://github.com/ClickHouse/ClickHouse/tree/master/docs) folder.

> If you are doing this for the first time, it's recommended to read the lightweight [Contributing to ClickHouse Documentation](https://github.com/ClickHouse/ClickHouse/tree/master/docs/README.md) guide first.


> Information about CI checks: https://clickhouse.tech/docs/en/development/continuous-integration/

pmed committed:
add aliases for `\<letter>` interactive client commands

initial implementation for:
  * \d aka SHOW TABLES
  * \l aka SHOW DATABASES
  * \c aka USE

Replace prefix of interactive `input` string with an aliased command
inside of the loop of `ClientBase::runInteractive()` in order to allow
command parameters, like `\c db_name`.

See issue #9339

alexey-milovidov commented:
> Automated testing seems not easy to do

Easy, we have tests with `expect` for this puspose :)

```
0_stateless$ ls -l *.expect
```

alexey-milovidov commented on src/Client/ClientBase.cpp:
```diff
@@ -1356,6 +1364,17 @@ void ClientBase::runInteractive()
 
+        for (const auto& [alias, command] : backslash_aliases)
+        {
+            if (input.starts_with(alias))
```
I've checked that `psql` supports these escape sequences even if you put extra whitespace before:

```
$ sudo -u postgres psql tutorial
psql (13.4 (Ubuntu 13.4-4.pgdg18.04+1))
Type "help" for help.

tutorial=#  \d
                List of relations
 Schema |         Name         | Type  |  Owner   
--------+----------------------+-------+----------
 public | hits_100m_obfuscated | table | postgres
(1 row)

tutorial=#        \d
                List of relations
 Schema |         Name         | Type  |  Owner   
--------+----------------------+-------+----------
 public | hits_100m_obfuscated | table | postgres
(1 row)
```

alexey-milovidov replied:
Also let's require whitespace separator to avoid accepting something like

`\da`

PS. I just noticed, that `\da` in PostgreSQL is listing aggregate functions :)

pmed replied:
Enhanced in 7b50ab7

pmed commented:
> > Automated testing seems not easy to do
> 
> Easy, we have tests with `expect` for this puspose :)
> 
> ```
> 0_stateless$ ls -l *.expect
> ```

Cool, thanks! I didn't know about `expect` program.

pmed committed:
skip spaces before and after \<letter> alias

... for proper single letter like `\l`, `\d` handling.

Using `std::search()` for alias substring search instead of `.starts_with()`
to get the alias start position, to check the leading range contains only space
characters, and to get the rest of input after the space.

See issue #9339

pmed commented on src/Client/ClientBase.cpp:
```diff
@@ -1366,12 +1366,19 @@ void ClientBase::runInteractive()
         {
-            if (input.starts_with(alias))
+            auto it = std::search(input.begin(), input.end(), alias.begin(), alias.end());
+            if (it != input.end() && std::all_of(input.begin(), it, isWhitespaceASCII))
```
In `ClientBase::processQueryText()` we have already trimmed text, so it may have sense to  process that trimmed string with.

pmed committed:
add test case for backslash letter commands

Using `expect` script because `\letter` aliases are
implemented only in interactive client mode.

vdimir commented:
Some issues in multyquery multiline mode (`clickhouse-client -mn`):

<details>
<summary>Details</summary>

```
-- empty output:
:) \d; 

-- ok:
:) \d ;

SHOW TABLES

Query id: e6563d65-3a48-4b23-827d-901d6b2fc45a

┌─name──┐
│ left  │
│ right │
└───────┘

2 rows in set. Elapsed: 0.019 sec. 

-- only first command executed:
:) \d ; \l ;

SHOW TABLES

Query id: 71a89571-1625-485e-802b-1d2f874888b3

┌─name──┐
│ left  │
│ right │
└───────┘

2 rows in set. Elapsed: 0.018 sec. 

:) \d ; \l;

SHOW TABLES

Query id: e153d191-312f-4af3-af25-5af19a86591e

┌─name──┐
│ left  │
│ right │
└───────┘

2 rows in set. Elapsed: 0.018 sec. 

-- empty output:
:) \d; \l;
:) 
```

</details>

alexey-milovidov commented:
It should work without semicolon even in multiquery mode.

pmed commented:
These backslash aliases are being processed on the input line processing, at the same level as the `quit` aliases. So they also don't work in the multi-query mode:

> % ./programs/clickhouse-client -n 
> :) quit; select 1; select 2; 
> 
> Syntax error: failed at position 1 ('quit'):
> 

>  % ./programs/clickhouse-client -mn
> :) select 1
> :-] ; select 2; quit
> :-] ;
> ... <skipped select results>
> Syntax error: failed at position 1 ('quit') (line 1, col 1):
> 
> quit
> ;
> 

As far as I understand, it can be resolved properly at the parsing. Another way with text replacements in the input string  would not work for quires with escaped characters, i.e. `select name from system.users where name like '\d'`

alexey-milovidov commented:
Sorry, I mean "multiline" mode :)

alexey-milovidov reviewed (commented):
LGTM
